test/integ/file_upload.cpp

```diff
@@ -0,0 +1,491 @@
+    struct stat file_info;
+    FILE *fd;
+
+    fd = fopen(TEST_CONTENT_FILEPATH, "rb");
+    if (!fd) {
+        return false;
+    }
+
+    if (fstat(fileno(fd), &file_info) != 0) {
+        return false;
+    }
+
+    curl = curl_easy_init();
+    if (!curl) {
+        fclose(fd);
+        return false;
+    }
+
+    curl_easy_setopt(curl, CURLOPT_URL, "localhost:8080/upload");
+    curl_easy_setopt(curl, CURLOPT_UPLOAD, 1L);
+    curl_easy_setopt(curl, CURLOPT_READDATA, fd);
+    curl_easy_setopt(curl, CURLOPT_INFILESIZE_LARGE, (curl_off_t)file_info.st_size);
+
+    res = curl_easy_perform(curl);
+    curl_easy_cleanup(curl);
+
+    fclose(fd);
+    if (res == CURLE_OK) {
+        return true;
+    }
+    return false;
+}
+
+class print_file_upload_resource : public http_resource {
+ public:
+     const shared_ptr<http_response> render_POST(const http_request& req) {
+         content = req.get_content();
+         args = req.get_args();
+         files = req.get_files();
+         shared_ptr<string_response> hresp(new string_response("OK", 201, "text/plain"));
+         return hresp;
+     }
+     const shared_ptr<http_response> render_PUT(const http_request& req) {
+         content = req.get_content();
+         args = req.get_args();
+         files = req.get_files();
+         shared_ptr<string_response> hresp(new string_response("OK", 200, "text/plain"));
+         return hresp;
+     }
+
+     const map<string, string, httpserver::http::arg_comparator> get_args() const {
+         return args;
+     }
+
+     const map<string, map<string, httpserver::http::file_info>> get_files() const {
+          return files;
+     }
+
+     const string get_content() const {
+         return content;
+     }
+
+ private:
+     map <string, string, httpserver::http::arg_comparator> args;
+     map<string, map<string, httpserver::http::file_info>> files;
+     string content;
+};
+
+LT_BEGIN_SUITE(file_upload_suite)
+    void set_up() {
+    }
+
+    void tear_down() {
+    }
+LT_END_SUITE(file_upload_suite)
+
+LT_BEGIN_AUTO_TEST(file_upload_suite, file_upload_memory_and_disk)
+    string upload_directory = ".";
+    webserver* ws;
+
+    ws = new webserver(create_webserver(8080)
+                       .put_processed_data_to_content()
+                       .file_upload_target(httpserver::FILE_UPLOAD_MEMORY_AND_DISK)
+                       .file_upload_dir(upload_directory)
+                       .generate_random_filename_on_upload());
+    ws->start(false);
+    LT_CHECK_EQ(ws->is_running(), true);
+
+    print_file_upload_resource resource;
+    ws->register_resource("upload", &resource);
+
+    CURLcode res = send_file_to_webserver(false, false);
+    LT_ASSERT_EQ(res, 0);
+
+    string actual_content = resource.get_content();
+    LT_CHECK_EQ(actual_content.find(FILENAME_IN_GET_CONTENT) != string::npos, true);
+    LT_CHECK_EQ(actual_content.find(TEST_CONTENT) != string::npos, true);
+
+    auto args = resource.get_args();
```

Comment:
nit: let's avoid `auto` in this context - it makes it harder to understand the type used (same for all the others in this code).